drivers/serial/uart_stm32.c

```diff
@@ -525,21 +525,32 @@ static void uart_stm32_poll_out(const struct device *dev,
 					unsigned char c)
 {
 	USART_TypeDef *UartInstance = UART_STRUCT(dev);
+#ifdef CONFIG_PM
 	struct uart_stm32_data *data = DEV_DATA(dev);
+#endif
+	int key;
 
-	/* Wait for TXE flag to be raised */
+	/* Wait for TXE flag to be raised
+	 * When TXE flag is raised, we lock interrupts to prevent interrupts (notably that of usart)
+	 * or thread switch. Then, we can safely send our character. The character sent will be
+	 * interlaced with the characters potentially send with interrupt transmission API
+	 */
 	while (1) {
-		if (atomic_cas(&data->tx_lock, 0, 1)) {
+		if (LL_USART_IsActiveFlag_TXE(UartInstance)) {
+			key = irq_lock();
 			if (LL_USART_IsActiveFlag_TXE(UartInstance)) {
 				break;
 			}
-			atomic_set(&data->tx_lock, 0);
+			irq_unlock(key);
```

Comment:
if (!k_is_in_isr() && key == 0){			
			k_yield();//yield execution to another thread of the same or higher priority. Its better for software, but not output pretty (some symbols  from shell some from kprint)
		}

Reply:
ok, k_yield could be interesting to reduce some cpu usage. But, I don't understand why you want check key?

Reply:
If key == 0 then  interrupts not disabled. So we can safe yield. I do not study  yield, when interrupts is disabled.